corehq/apps/user_importer/importer.py

```diff
@@ -457,131 +463,89 @@ def create_or_update_users_and_groups(upload_domain, user_specs, upload_user, gr
-                for key in dict(data, **uncategorized_data):
-                    value = user.metadata[key]
-                    if value is None or value == '':
-                        user.pop_metadata(key)
+                    commcare_user_importer.update_name(name)
+
+                commcare_user_importer.update_user_data(data, uncategorized_data, profile, domain_info)
 
                 if language:
-                    user.language = language
+                    commcare_user_importer.update_language(language)
                 if email:
-                    user.email = email.lower()
+                    commcare_user_importer.update_email(email)
                 if is_active is not None:
-                    user.is_active = is_active
+                    commcare_user_importer.update_status(is_active)
 
+                # Do this here so that we validate the location code before we
+                # save any other information to the user, this way either all of
+                # the user's information is updated, or none of it
+                # Do not update location info if the column is not included at all
                 if domain_info.can_assign_locations and location_codes is not None:
-                    # Do this here so that we validate the location code before we
-                    # save any other information to the user, this way either all of
-                    # the user's information is updated, or none of it
-
-                    # Do not update location info if the column is not included at all
-                    location_ids = find_location_id(location_codes, domain_info.location_cache)
-                    locations_updated, primary_loc_removed = check_modified_user_loc(location_ids,
-                                                                                     user.location_id,
-                                                                                     user.assigned_location_ids)
-                    if primary_loc_removed:
-                        user.unset_location(commit=False)
-                    if locations_updated:
-                        user.reset_locations(location_ids, commit=False)
+                    commcare_user_importer.update_locations(location_codes, domain_info)
 
                 if role:
                     role_qualified_id = domain_info.roles_by_name[role]
-                    user_current_role = user.get_role(domain=domain)
-                    log_role_update = not (user_current_role
-                                        and user_current_role.get_qualified_id() == role_qualified_id)
-                    if log_role_update:
-                        user.set_role(domain, role_qualified_id)
-
-                if web_user:
-                    user.update_metadata({'login_as_user': web_user})
+                    commcare_user_importer.update_role(role, domain_info)
```

Comment:
why not pass in `role_qualified_id` here so that this matches the `WebUserImporter.update_role` implementation. I think would then have exactly the same implementation and could be put in a super class maybe?

Reply:
thank you. role_qualified_id was already available, so we can just pass that itself

https://github.com/dimagi/commcare-hq/pull/29935/commits/2318dd38c185b16a60d95cfd3a05acaa91e3ddb1

and
https://github.com/dimagi/commcare-hq/pull/29935/commits/d4c56fd788915b15bef29ba535a4e90eb9620144